cluster_experiments/experiment_analysis.py

```diff
@@ -680,3 +683,186 @@ def analysis_point_estimate(self, df: pd.DataFrame, verbose: bool = False) -> fl
+
+    target_values = np.random.normal(0, 1, size=len(combinations))
+
+    df = pd.DataFrame(combinations, columns=["user", "date"])
+    df["target"] = target_values
+
+    df["treatment_period"] = np.where(
+        df["date"].dt.date < date(2022, 1, 15), "Before", "After"
+    )
+
+    SyntheticControlAnalysis(
+        cluster_cols=["user"], time_col="date", transition_date=date(2022, 1, 15)
+    ).get_pvalue(df)
+
+    ```
+    """
+
+    def __init__(
+        self,
+        target_col: str = "target",
+        treatment_col: str = "treatment",
+        treatment: str = "B",
+        cluster_cols: Optional[List[str]] = None,
+        hypothesis: str = "two-sided",
+        time_col: str = "date",
+        transition_date: str = None,
+    ):
+        super().__init__(cluster_cols)
+
+        self.target_col = target_col
+        self.treatment = treatment
+        self.treatment_col = treatment_col
+        self.cluster_cols = cluster_cols
+        self.hypothesis = hypothesis
+        self.time_col = time_col
+        self.transition_date = transition_date
+
+    def fit_synthetic(self, df: pd.DataFrame) -> list:
+        """Returns the fitted OLS model"""
+
+        # todo should I throw an error if more than one cluster column is passed?
+
+        X = (
+            df.query(f"{self.treatment_col}==0")
+            .pivot(index=self.cluster_cols, columns=self.time_col)[self.target_col]
+            .T
+        )
+
+        if not any(df[self.treatment_col] == 1):
+            raise ValueError("No treatment unit found in the data.")
+
+        y = df.query(f"{self.treatment_col}==1")[f"{self.target_col}"]  # treated
+
+        weights = get_w(X, y)
+
+        return weights
+
+    def synthetic_control(self, df: pd.DataFrame, treatment_unit) -> np.array:
+        weights = self.fit_synthetic(df)
+
+        synthetic = (
+            df.query(
+                f"{''.join(self.cluster_cols)} != '{treatment_unit}'"
+            )  # remove treatment cluster
+            .pivot(index=self.time_col, columns=self.cluster_cols)[self.target_col]
+            .values.dot(weights)
+        )
+
+        return df.query(
+            f"{''.join(self.cluster_cols)} == '{treatment_unit}'"
+        ).assign(  # add synthetic to treatment unit
+            synthetic=synthetic
+        )
+
+    def pvalue_based_on_hypothesis(
+        self, synthetic_donors: pd.DataFrame, treatment_cluster: str
+    ) -> float:
+
+        treatment_index = [
+            i
+            for i, df in enumerate(synthetic_donors)
+            if (df[self.treatment_col] == 1).any()
+        ]
+
+        synthetic_donors = [
+            unit.assign(effect=unit[self.target_col] - unit["synthetic"])
+            for unit in synthetic_donors
+        ]
+
+        # first calculate the average effect after intervention for each unit, in a list comprehension
+        avg_effects = [
+            unit.query("treatment_period == 'After'")["effect"].mean()
```

Comment:
I think it makes sense to add 'After' as an attribute of the class instead of being hardcoded